File: 0876 - Middle of the Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        new_head = head
        length = 0
        while new_head:
            length += 1
            new_head = new_head.next
        second_half = length // 2
        new_head = head
        i = 0
        while i < second_half:
            new_head = new_head.next
            i += 1
        return  new_head

File: 0876 - Middle of the Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        int length = 0;
        struct ListNode* new_head = head;
        while(new_head){new_head = new_head->next; length++;}
        length >>= 1;
        new_head = head;
        int i = 0;
        while(length--){new_head = new_head->next;}
        return new_head;
        }
};